### PR TITLE
Kddimitrov/livesync tool tests

### PR DESCRIPTION
**Pull request by KristianDD**
Add tests for android-livesync-tool

### CODE REVIEW

File: test/services/livesync/android-livesync-tool.ts [hunk 1]
```diff
@@ -0,0 +1,541 @@
...
+
+		if (cb && cb instanceof Function) {
+			cb.call(this);
+		}
+
+		return true;
+	}
+}
+
+const rootTestFileJs = "test.js";
```
[the remark this answers is not in the document]
Comment (KristianDD) [reply test/services/livesync/android-livesync-tool.ts]: The idea is that this is a root file in the app folder. Its actually a path relative to the app folder. Will try to rename it accordingly.

File: test/services/livesync/android-livesync-tool.ts [hunk 2]
```diff
@@ -0,0 +1,541 @@
...
+const rootTestFileCss = "test.css";
+const dirTestFileJs = path.join("testdir", "testdir.js");
```
[the remark this answers is not in the document]
Comment (KristianDD) [reply test/services/livesync/android-livesync-tool.ts]: A file one level deeper in app folder. Will try to rename it accordingly.

File: test/services/livesync/android-livesync-tool.ts [hunk 3]
```diff
@@ -0,0 +1,541 @@
...
+	beforeEach(() => {
+		sandbox = sinon.sandbox.create();
+		testSocket = new TestSocket();
+		fileStreams = {};
+		testInjector = createTestInjector(testSocket, fileStreams);
+		createTestProject(testInjector);
+		livesyncTool = testInjector.resolve(AndroidLivesyncTool);
+	});
+
+	afterEach(() => {
```
[the remark this answers is not in the document]
Comment (KristianDD) [reply test/services/livesync/android-livesync-tool.ts]: I think that every time we create new instance of `livesyncTool` in `beforeEach`. We should be safe.

File: test/services/livesync/android-livesync-tool.ts [hunk 4]
```diff
@@ -0,0 +1,541 @@
...
+
+	describe("methods", () => {
+		describe("connect", () => {
+			it("should retry if first connect fails", () => {
+				//arrange
+				const originalOn = testSocket.on;
+				const originalOnce = testSocket.once;
+				const connectStub: sinon.SinonStub = sandbox.stub(testSocket, "connect");
+				let dataAttachCount = 0;
+				let closeAttachCount = 0;
```
[the remark this answers is not in the document]
Comment (KristianDD) [reply test/services/livesync/android-livesync-tool.ts]: It counts not the time the data event is called, but how many times someone attached to the event.

[the remark this answers is not in the document]
Comment (Fatme) [reply test/services/livesync/android-livesync-tool.ts]: Oo yes, my bad... `closeAttachCount ` and `dataAttachCount` are perfect.

File: test/services/livesync/android-livesync-tool.ts [hunk 5]
```diff
@@ -0,0 +1,533 @@
...
+		livesyncTool = testInjector.resolve(AndroidLivesyncTool);
+	});
+
+	afterEach(() => {
+		sandbox.restore();
+	});
+
+	describe("methods", () => {
+		describe("connect", () => {
+			it("should retry if first socket connect emitts close", async () => {
```
Comment (Fatme) [inline test/services/livesync/android-livesync-tool.ts]: Typo? `emits`?

File: test/services/livesync/android-livesync-tool.ts [hunk 6]
```diff
@@ -0,0 +1,533 @@
...
+
+				//act
+				await livesyncTool.connect(connectData);
+
+				//assert
+				assert(connectStub.calledTwice);
+				assert.equal(livesyncTool.protocolVersion, protocolVersion);
+			});
+
+			it("should retry if first socket connect erros", () => {
```
Comment (Fatme) [inline test/services/livesync/android-livesync-tool.ts]: Typo? `errors`

File: test/services/livesync/android-livesync-tool.ts [hunk 7]
```diff
@@ -0,0 +1,533 @@
...
+				});
+			});
+
+			describe("sendDoSync", () => {
+				it("resolves after received data", () => {
+					//arrange
+					let doSyncResolved = false;
+
+					//act
+					const doSyncPromise = livesyncTool.sendDoSyncOperation(true);
```
Comment (Fatme) [inline test/services/livesync/android-livesync-tool.ts]: This is more readable 
```
livesyncTool.sendDoSyncOperation({ doRefresh: true })
```

File: test/services/livesync/android-livesync-tool.ts [hunk 8]
```diff
@@ -0,0 +1,533 @@
...
+					const doSyncData = getSyncData((testSocket as TestSocket).accomulatedData);
+					doSyncPromise.then(() => {
+						doSyncResolved = true;
+					}).catch(assert.fail);
+
+					//assert
+					assert.isFalse(doSyncResolved);
+					testSocket.emit('data', getSyncResponse(AndroidLivesyncTool.OPERATION_END_NO_REFRESH_REPORT_CODE, doSyncData.operationUid));
+
+					return doSyncPromise.then(function() {
```
Comment (Fatme) [inline test/services/livesync/android-livesync-tool.ts]:
```
doSyncPromise.then(() => {
```

File: test/services/livesync/android-livesync-tool.ts (hunk 7 above)
Comment (KristianDD) [reply test/services/livesync/android-livesync-tool.ts]: I think we have discussed this already. Having a definition of the method shows exactly how each parameter is called. If you still insist I can change it.

**Pull request merged**